### PR TITLE
JavaScript support

### PR DESCRIPTION
**Pull request by Maximilian-Seitz**

### CODE REVIEW

File: javascript/js-parser/build.gradle [hunk 1]
```diff
@@ -0,0 +1,9 @@
+plugins {
+    id 'kotlin'
+}
+
+
+dependencies {
+    //https://mvnrepository.com/artifact/org.graalvm.js/js
+    compile group: 'org.graalvm.js', name: 'js', version: gradle.graaljsVersion
```
Comment (Schahen) [inline javascript/js-parser/build.gradle]: here's the way we actually add compile dependencies:
`implementation("org.jetbrains.kotlin:kotlin-stdlib:$gradle.kotlinVersion")`

File: javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    for (line in scriptBody.statements) {
+        println(line.lineNumber.toString() + " <" + line.javaClass.name + ">\t" + line)
+
+        if (line is ExpressionStatement) {
+            val expression = line.expression
+
+            if (expression.isAssignment) {
+
+                if (expression is BinaryNode) {
+                    println("\t" + expression.assignmentDest)
```
Comment (Schahen) [inline javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt]: check out
``` const val FORMAT_TAB = "    "```

File: javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    val source = Source.sourceFor(fileName, sourceCode)
+
+    val parser = Parser(ScriptEnvironment.builder().ecmaScriptVersion(Integer.MAX_VALUE).build(), source, ErrorManager.ThrowErrorManager())
+
+    val baseNode = parser.parse()
+
+    val scriptBody = baseNode.body
+
+    for (line in scriptBody.statements) {
+        println(line.lineNumber.toString() + " <" + line.javaClass.name + ">\t" + line)
```
Comment (Schahen) [inline javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt]: whenever we are talking about the code ready to be in prod let's not have printlns

File: javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import com.oracle.js.parser.ErrorManager
+import com.oracle.js.parser.Parser
+import com.oracle.js.parser.ScriptEnvironment
+import com.oracle.js.parser.Source
+import com.oracle.js.parser.ir.*
+import java.io.File
+
+
+fun parseJS(fileName: String) {
+    val sourceCode = File(fileName).readText()
```
Comment (Schahen) [inline javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt]: What's the purpose of leaving empty line after each new declaration?

File: javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package org.jetbrains.dukat.jsParser
+
+import com.oracle.js.parser.ErrorManager
+import com.oracle.js.parser.Parser
+import com.oracle.js.parser.ScriptEnvironment
+import com.oracle.js.parser.Source
+import com.oracle.js.parser.ir.*
```
Comment (Schahen) [inline javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt]: it's a star import

File: javascript/js-lowerings/src/org/jetbrains/dukat/js/lowerings/lowerings.kt [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+                abstract = false,
+                visibilityModifier = VisibilityModifier.DEFAULT
+        )
+    }
+
+    private fun JSDeclaration.convert(): TopLevelModel {
+        return when(this) {
+            is JSFunctionDeclaration -> this.convert()
+            is JSClassDeclaration -> this.convert()
+            is JSReferenceDeclaration -> throw IllegalStateException("References need to be resolved before conversion to model.")
```
[the remark this answers is not in the document]
Comment (Maximilian-Seitz) [reply javascript/js-lowerings/src/org/jetbrains/dukat/js/lowerings/lowerings.kt]: I thought the same, so I'm doing that now.

File: javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt (hunk 5 above)
Comment (Maximilian-Seitz) [reply javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt]: I resolved those

File: javascript/js-parser/build.gradle (hunk 1 above)
Comment (Maximilian-Seitz) [reply javascript/js-parser/build.gradle]: I changed that now

File: javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt (hunk 3 above)
Comment (Maximilian-Seitz) [reply javascript/js-parser/src/org/jetbrains/dukat/jsParser/parseJS.kt]: I removed those

**Pull request closed without merging**